packages/enzyme-adapter-utils/src/Utils.js

```diff
@@ -267,18 +267,35 @@ export function simulateError(
   getNodeType = nodeTypeFromType,
   getDisplayName = displayNameOfNode,
 ) {
-  const { componentDidCatch } = catchingInstance || {};
-  if (!componentDidCatch) {
+  const instance = catchingInstance || {};
+
+  const { componentDidCatch } = instance;
+
+  const { getDerivedStateFromError } = instance.constructor;
+
+  if (!componentDidCatch && !getDerivedStateFromError) {
     throw error;
   }
 
-  const componentStack = getComponentStack(
-    hierarchy,
-    getNodeType,
-    getDisplayName,
-  );
+  function setDerivedStateFromError() {
+    const stateUpdate = getDerivedStateFromError.call(instance.constructor, error);
+    if (stateUpdate) {
+      instance.setState(stateUpdate);
+    }
+  }
 
-  componentDidCatch.call(catchingInstance, error, { componentStack });
+  function callComponentDidCatch() {
+    const componentStack = getComponentStack(hierarchy, getNodeType, getDisplayName);
+    componentDidCatch.call(catchingInstance, error, { componentStack });
+  }
+
+  if (getDerivedStateFromError) {
+    setDerivedStateFromError();
+  }
+
+  if (componentDidCatch) {
+    callComponentDidCatch();
```

Comment:
can you confirm that this is the order these get called by react itself?

Reply:
`ComponentDidCatch` is always called as a callback to state update from `getDerivedStateFromError`, so the order is the same. See [ReactFiberUndwindWork#createClassErrorUpdate](https://github.com/facebook/react/blob/1e3b6192b54df20ac117a2af56afbe00ac9487b7/packages/react-reconciler/src/ReactFiberUnwindWork.js#L103) for comparison. 
However, I am not sure if calling `instance.setState` with state update from `gDSFE` is an accurate simulation. Do you have an idea how to make sure that `componentDidCatch` has access to the freshly updated state?

Reply:
Hmm - let's start by setting up very explicit test cases. We already have some that show that cDC is called, with the right info; we now have some that gDSFE is called, with the right info - now we need to ensure that cDC is called (with the right info) both when gDFSE returns new state, and also when it does not return changed state.

Once we get those set up (verified against actual React behavior), then it'll be easier to decide how to implement the behavior :-)

Reply:
So actually React first tries to rerender after gDSFE and it fails when state remains the same, so cDC is never called. I pushed tests for this and have also a small codesandbox with such behaviour (https://codesandbox.io/s/q7r2z35y4w?fontsize=14).